hal/api/Event.h

```diff
@@ -0,0 +1,397 @@
+     */
+    template <typename T>
+    void attach(volatile T *obj,  void (*func)(volatile T*), EventLoop *queue = 0);
+
+    /** Attach a static function with a bound pointer
+     *  @param obj  Pointer to object to bind to function
+     *  @param func Static function to attach
+     *  @param queue Pointer to the event queue (NULL to call the event immediately)
+     */
+    template <typename T>
+    void attach(const volatile T *obj,  void (*func)(const volatile T*), EventLoop *queue = 0);
+
+    /** Attach a member function
+     *  @param obj  Pointer to object to invoke member function on
+     *  @param func Member function to attach
+     *  @param queue Pointer to the event queue (NULL to call the event immediately)
+     */
+    template<typename T>
+    void attach(T *obj,  void (T::*func)(), EventLoop *queue = 0);
+
+    /** Attach a member function
+     *  @param obj  Pointer to object to invoke member function on
+     *  @param func Member function to attach
+     *  @param queue Pointer to the event queue (NULL to call the event immediately)
+     */
+    template<typename T>
+    void attach(const T *obj,  void (T::*func)() const, EventLoop *queue = 0);
+
+    /** Attach a member function
+     *  @param obj  Pointer to object to invoke member function on
+     *  @param func Member function to attach
+     *  @param queue Pointer to the event queue (NULL to call the event immediately)
+     */
+    template<typename T>
+    void attach(volatile T *obj,  void (T::*func)() volatile, EventLoop *queue = 0);
+
+    /** Attach a member function
+     *  @param obj  Pointer to object to invoke member function on
+     *  @param func Member function to attach
+     *  @param queue Pointer to the event queue (NULL to call the event immediately)
+     */
+    template<typename T>
+    void attach(const volatile T *obj,  void (T::*func)() const volatile, EventLoop *queue = 0);
+
+    /** Calls the event. Depending on the call mode, the callback is either
+     *  executed immediately (if _queue is NULL) or enqueue in _queue (if
+     *  _queue is not NULL).
+     */
+    void call() const;
+
+    /** Calls the event. Depending on the call mode, the callback is either
+     *  executed immediately (if _queue is NULL) or enqueue in _queue (if
+     *  _queue is not NULL).
+     */
+    void operator()() const;
+
+    /** Static thunk for passing as C-style function
+     *  @param func Event to call passed as void pointer
+     */
+    static void thunk(void *func);
+
+    /** Check if the callback instance is attached to a function
+     */
+    operator bool() const;
+
+private:
+    Callback<void()> _callback;
+    EventLoop *_queue;
+};
+
+/** Create a callback class with type infered from the arguments
+ *
+ *  @param func Static function to attach
+ *  @param queue Pointer to the event queue
+ *  @return     Callback with infered type
+ */
+Event event(void (*func)(void), EventLoop *queue);
```

Comment:
This may be subjective, but we could provide these convenience functions as member functions on the `EventLoop` class:
``` cpp
ticker.attach(queue->event(func));
// vs
ticker.attach(event(func, &queue));
```

This avoids ambiguity on if `queue` effects the event or is just the argument to `func`.